app/services/combined_audio_derivative_creator.rb

```diff
@@ -0,0 +1,170 @@
+require 'tempfile'
+
+
+Response = Struct.new(:webm_file, :mp3_file, :fingerprint, :start_times, keyword_init: true)
+#
+# Generate the combined audio derivatives for a
+# given oral history work.
+#
+# some_work = Work.find_by_friendlier_id(friendlier_id)
+# CombinedAudioDerivativeCreator.new(some_work).generate
+#
+# Sample output:
+# <struct Response
+#  webm_file=<File:/path/to/file.webm>,
+#  mp3_file=<File:/path/to/file.mp3>,
+#  fingerprint="e0248c40015d6dc90b0d02937950b5d7",
+#  start_times=
+#   [["ddbd1a8d-c2eb-47b3-85a4-11b4ffb41719", 0],
+#    ["df773502-56d7-4756-a58c-b1f479910e97", 1.593469]]
+#  >
+
+class CombinedAudioDerivativeCreator
+
+  attr_reader :work
+
+  def initialize(work)
+    @cmd = TTY::Command.new(printer: :null)
+    @work = work
+  end
+
+  def generate
+    output_files = {}
+
+    # Create the two output files:
+    ['mp3', 'webm'].each do |format|
+      output_files[format] = output_file(format)
+      ffmpeg_args = args_for_ffmpeg(output_files[format].path)
+      log "Creating #{format} using: #{ffmpeg_args.join(" ")}"
+      @cmd.run(*ffmpeg_args, binmode: true)
+    end
+
+    resp = Response.new
+    resp.webm_file   = output_files['webm']
+    resp.mp3_file    = output_files['mp3']
+    resp.fingerprint = fingerprint
+    resp.start_times = calculate_start_times
+
+    # Before leaving, get rid of the downloaded originals:
+    components.map!(&:unlink)
+
+    resp
+  end
+
+  def output_file(format)
+    Tempfile.new(['output', ".#{format}"], :encoding => 'binary')
+  end
+
+  # Use ffprobe to determine the length of an audio file.
+  def duration_of_audio_file_in_seconds(path)
+    options = ['ffprobe', '-v', 'error',
+      '-show_entries', 'format=duration', '-of',
+      'default=noprint_wrappers=1:nokey=1'
+    ] + [ path ]
+    @cmd.run(*options).out.strip.to_f
+  end
+
+  def components
+    @components ||= begin
+      result = []
+      audio_member_files.each do |original_file|
+        log "Downloading #{original_file.metadata['filename']}"
+        new_temp_file = Tempfile.new(['temp_', original_file.metadata['filename'].downcase], :encoding => 'binary')
+        original_file.open(rewindable:false) do |input_audio_io|
+          new_temp_file.write input_audio_io.read until input_audio_io.eof?
+        end
+
+        log "Finished downloading #{original_file.metadata['filename']}"
+        result << new_temp_file
+      end
+      result
+    end
+  end
+
+  # A list of arrays; the first item in each is the UUID of each audio member,
+  # while the second is the *starting point* of that audio w/r/t the combined audio.
+  def calculate_start_times()
+    durations = components.map do |f|
+      duration_of_audio_file_in_seconds(f.path)
```

Comment:
This is going to do a lot of extra shell-outs to `ffmpeg`. When you ask for start time for file 2, it'll shell out to `ffmpeg` for 1. When you ask for 3, it'll shell out for 1 and 2 (again). Etc. 

Because the info is never cached, every time you ask for it it shells out. Is there another way to do this to avoid that? Or does it not really matter?

Reply:
We cleared this up on Slack; resolving. (TLDR: no, there are no extra shell outs.)